airbyte-cdk/python/airbyte_cdk/sources/file_based/stream/default_file_based_stream.py

```diff
@@ -3,37 +3,62 @@
 #
 
 import asyncio
+import itertools
 import logging
-from datetime import datetime
+from datetime import timedelta
 from functools import cache
-from typing import Any, Iterable, List, Mapping, Optional, Union
+from typing import Any, Iterable, List, Mapping, MutableMapping, Optional, Union
 
-from airbyte_cdk.models import SyncMode
-from airbyte_cdk.sources.declarative.types import StreamSlice, StreamState
+from airbyte_cdk.sources.declarative.types import StreamSlice
 from airbyte_cdk.sources.file_based.exceptions import MissingSchemaError, RecordParseError, SchemaInferenceError
 from airbyte_cdk.sources.file_based.remote_file import RemoteFile
-from airbyte_cdk.sources.file_based.schema_helpers import merge_schemas, type_mapping_to_jsonschema
+from airbyte_cdk.sources.file_based.schema_helpers import merge_schemas
 from airbyte_cdk.sources.file_based.schema_validation_policies import record_passes_validation_policy
 from airbyte_cdk.sources.file_based.stream import AbstractFileBasedStream
+from airbyte_cdk.sources.file_based.stream.file_based_cursor import FileBasedCursor
+from airbyte_cdk.sources.streams import IncrementalMixin
 from airbyte_cdk.sources.utils.record_helper import stream_data_to_airbyte_message
 
 
-class DefaultFileBasedStream(AbstractFileBasedStream):
+class DefaultFileBasedStream(AbstractFileBasedStream, IncrementalMixin):
+
     """
     The default file-based stream.
     """
 
+    ab_last_mod_col = "_ab_source_file_last_modified"
+    ab_file_name_col = "_ab_source_file_url"
+    airbyte_columns = [ab_last_mod_col, ab_file_name_col]
+
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+        self._cursor = FileBasedCursor(
+            self.config.max_history_size or 10_000, timedelta(days=(self.config.days_to_sync_if_history_is_full or 3)), self.logger
+        )
+
+    @property
+    def state(self) -> MutableMapping[str, Any]:
+        return self._cursor.get_state()
+
+    @state.setter
+    def state(self, value: MutableMapping[str, Any]):
+        """State setter, accept state serialized by state getter."""
+        self._cursor.set_initial_state(value)
+
     @property
     def primary_key(self) -> Optional[Union[str, List[str], List[List[str]]]]:
         return self.config.primary_key
 
-    def read_records(
-        self,
-        sync_mode: SyncMode,
-        cursor_field: List[str] = None,
-        stream_slice: Optional[StreamSlice] = None,
-        stream_state: Optional[StreamState] = None,
-    ) -> Iterable[Mapping[str, Any]]:
+    def compute_slices(self) -> Iterable[Optional[Mapping[str, Any]]]:
+        # Group all files by timestamps and return them as slices
+        files = [{"uri": f.uri, "last_modified": f.last_modified, "file_type": f.file_type} for f in self.list_files_for_this_sync()]
+
+        slices = [{"files": list(group[1])} for group in itertools.groupby(files, lambda f: f["last_modified"])]
```

Comment:
Why are we grouping files on a specific time stamp? Is it because it's harder to track state? I just want to acknowledge the downside which is when we will want to parallelize, we might have less benefits (given it's frequent that files are modified at the exact same time)

Reply:
Yes, it's about state management.

One slice per file would trivially work if we could always keep the history in the state, but we need an ordering guarantee since files can be removed from the history.

We can still sync files in parallel by parallelizing [this block within the read_records](https://github.com/airbytehq/airbyte/blob/master/airbyte-cdk/python/airbyte_cdk/sources/file_based/stream/default_file_based_stream.py#L45-L53). We just cannot commit them until they're all synced